File: sdk/python/streaming.py
"""Streaming utilities for the SuperDev Python SDK."""

from __future__ import annotations

from typing import Any, Callable

from sdk.python.types import StreamingChunk


class StreamProcessor:
    """Processes streaming chunks from the SuperDev API.

    Example::

        processor = StreamProcessor()
        processor.on_chunk(lambda chunk: print(chunk.delta, end=""))
        processor.on_complete(lambda usage: print(f"\\nTokens: {usage}"))
        for chunk in client.chat.stream("Hello"):
            processor.process(chunk)
    """

    def __init__(self) -> None:
        self._on_chunk: Callable[[StreamingChunk], None] | None = None
        self._on_complete: Callable[[dict[str, int]], None] | None = None
        self._on_error: Callable[[Exception], None] | None = None
        self._buffer: list[str] = []
        self._usage: dict[str, int] = {}

    def on_chunk(self, callback: Callable[[StreamingChunk], None]) -> StreamProcessor:
        self._on_chunk = callback
        return self

    def on_complete(self, callback: Callable[[dict[str, int]], None]) -> StreamProcessor:
        self._on_complete = callback
        return self

    def on_error(self, callback: Callable[[Exception], None]) -> StreamProcessor:
        self._on_error = callback
        return self

    def process(self, chunk: StreamingChunk) -> None:
        if chunk.delta:
            self._buffer.append(chunk.delta)
        if chunk.usage:
            self._usage = chunk.usage
        if self._on_chunk:
            self._on_chunk(chunk)
        if chunk.finish_reason and self._on_complete:
            self._on_complete(self._usage)

    @property
    def full_text(self) -> str:
        return "".join(self._buffer)

    @property
    def usage(self) -> dict[str, int]:
        return self._usage

    def reset(self) -> None:
        self._buffer.clear()
        self._usage = {}


class StreamBuffer:
    """Buffers streaming output for reassembly."""

    def __init__(self, max_size: int = 10_000) -> None:
        self._chunks: list[str] = []
        self._max_size = max_size
        self._total_chars = 0

    def add(self, delta: str) -> None:
        if self._total_chars + len(delta) > self._max_size:
            overflow = self._total_chars + len(delta) - self._max_size
            self._chunks = self._chunks[1:] if self._chunks else []
            self._total_chars = max(0, self._total_chars - overflow)
        self._chunks.append(delta)
        self._total_chars += len(delta)

    def get_text(self) -> str:
        return "".join(self._chunks)

    def clear(self) -> None:
        self._chunks.clear()
        self._total_chars = 0

    @property
    def size(self) -> int:
        return self._total_chars

File: sdk/typescript/src/streaming.ts
import type { StreamingChunk } from "./types";

/** Callback types for the stream processor. */
type ChunkCallback = (chunk: StreamingChunk) => void;
type CompleteCallback = (usage: Record<string, number>) => void;
type ErrorCallback = (error: Error) => void;

/**
 * Processes streaming chunks from the SuperDev API.
 *
 * @example
 * ```ts
 * const processor = new StreamProcessor();
 * processor
 *   .onChunk((chunk) => process.stdout.write(chunk.delta))
 *   .onComplete((usage) => console.log("\nTokens:", usage));
 *
 * for await (const chunk of client.chat.stream("Tell me a story")) {
 *   processor.process(chunk);
 * }
 * ```
 */
export class StreamProcessor {
  private onChunkCb: ChunkCallback | null = null;
  private onCompleteCb: CompleteCallback | null = null;
  private onErrorCb: ErrorCallback | null = null;
  private buffer: string[] = [];
  private usage: Record<string, number> = {};

  /** Register a callback for each incoming chunk. */
  onChunk(callback: ChunkCallback): this {
    this.onChunkCb = callback;
    return this;
  }

  /** Register a callback invoked when the stream completes. */
  onComplete(callback: CompleteCallback): this {
    this.onCompleteCb = callback;
    return this;
  }

  /** Register a callback for stream errors. */
  onError(callback: ErrorCallback): this {
    this.onErrorCb = callback;
    return this;
  }

  /** Process a single streaming chunk. */
  process(chunk: StreamingChunk): void {
    if (chunk.delta) {
      this.buffer.push(chunk.delta);
    }
    if (chunk.usage && Object.keys(chunk.usage).length > 0) {
      this.usage = { ...this.usage, ...chunk.usage };
    }
    if (this.onChunkCb) {
      this.onChunkCb(chunk);
    }
    if (chunk.finishReason && this.onCompleteCb) {
      this.onCompleteCb(this.usage);
    }
  }

  /** Process a stream error. */
  processError(error: Error): void {
    if (this.onErrorCb) {
      this.onErrorCb(error);
    }
  }

  /** The fully reassembled text from all processed chunks. */
  get fullText(): string {
    return this.buffer.join("");
  }

  /** The accumulated token usage. */
  get totalUsage(): Record<string, number> {
    return { ...this.usage };
  }

  /** Reset internal state. */
  reset(): void {
    this.buffer = [];
    this.usage = {};
  }
}

/**
 * Buffers streaming output for reassembly with a max size cap.
 */
export class StreamBuffer {
  private chunks: string[] = [];
  private totalChars = 0;
  private readonly maxSize: number;

  constructor(maxSize = 10_000) {
    this.maxSize = maxSize;
  }

  /** Append a delta string to the buffer. */
  add(delta: string): void {
    if (this.totalChars + delta.length > this.maxSize) {
      const overflow = this.totalChars + delta.length - this.maxSize;
      const removed = this.chunks.shift();
      if (removed) {
        this.totalChars -= removed.length;
      }
      this.totalChars = Math.max(0, this.totalChars - overflow);
    }
    this.chunks.push(delta);
    this.totalChars += delta.length;
  }

  /** Get the full buffered text. */
  getText(): string {
    return this.chunks.join("");
  }

  /** Clear the buffer. */
  clear(): void {
    this.chunks = [];
    this.totalChars = 0;
  }

  /** Current buffer size in characters. */
  get size(): number {
    return this.totalChars;
  }
}
